src/node_http_common.h

```diff
@@ -0,0 +1,521 @@
+// for an example.
+template <typename T>
+class NgHeaders {
+ public:
+  typedef typename T::nv_t nv_t;
+  inline NgHeaders(Environment* env, v8::Local<v8::Array> headers);
+  ~NgHeaders() = default;
+
+  const nv_t* operator*() const {
+    return reinterpret_cast<const nv_t*>(*buf_);
+  }
+
+  const nv_t* data() const {
+    return reinterpret_cast<const nv_t*>(*buf_);
+  }
+
+  size_t length() const {
+    return count_;
+  }
+
+ private:
+  size_t count_;
+  MaybeStackBuffer<char, 3000> buf_;
+};
+
+// The ng libraries (nghttp2 and nghttp3) each use nearly identical
+// reference counted structures for retaining header name and value
+// information in memory until the application is done with it.
+// The NgRcBufPointer is an intelligent pointer capable of working
+// with either type, handling the ref counting increment and
+// decrement as appropriate. The Template takes a single Traits
+// type that provides the rc buffer and vec type, as well as
+// implementations for multiple static functions.
+// See Http2RcBufferPointerTraits in node_http2.h for an example.
+template <typename T>
+class NgRcBufPointer : public MemoryRetainer {
+ public:
+  typedef typename T::rcbuf_t rcbuf_t;
+  typedef typename T::vector_t vector_t;
+  NgRcBufPointer() {}
+
+  void MemoryInfo(MemoryTracker* tracker) const override {
+    tracker->TrackFieldWithSize("buf", len(), "buf");
+  }
+
+  SET_MEMORY_INFO_NAME(NgRcBufPointer)
+  SET_SELF_SIZE(NgRcBufPointer)
+
+  explicit NgRcBufPointer(rcbuf_t* buf) {
+    reset(buf);
+  }
+
+  template <typename B>
+  NgRcBufPointer(const NgRcBufPointer<B>& other) {
+    reset(other.get());
+  }
+
+  NgRcBufPointer(const NgRcBufPointer& other) {
+    reset(other.get());
+  }
+
+  template <typename B>
+  NgRcBufPointer& operator=(const NgRcBufPointer<B>& other) {
+    if (other.get() == get()) return *this;
+    this->~NgRcBufPointer();
+    return *new (this) NgRcBufPointer(other);
+  }
+
+  NgRcBufPointer& operator=(const NgRcBufPointer& other) {
+    if (other.get() == get()) return *this;
+    this->~NgRcBufPointer();
+    return *new (this) NgRcBufPointer(other);
+  }
+
+  NgRcBufPointer(NgRcBufPointer&& other) {
+    this->~NgRcBufPointer();
+    buf_ = other.buf_;
+    other.buf_ = nullptr;
+  }
+
+  NgRcBufPointer& operator=(NgRcBufPointer&& other) {
+    this->~NgRcBufPointer();
+    return *new (this) NgRcBufPointer(std::move(other));
+  }
+
+  ~NgRcBufPointer() {
+    rcbuf_t* ptr = get();
+    if (ptr != nullptr) {
+      T::dec(ptr);
+    }
+  }
+
+  // Returns the underlying ngvec for this rcbuf
+  uint8_t* data() const {
+    vector_t v = T::get_vec(buf_);
+    return v.base;
+  }
+
+  size_t len() const {
+    vector_t v = T::get_vec(buf_);
+    return v.len;
+  }
+
+  std::string str() const {
+    return std::string(reinterpret_cast<const char*>(data()), len());
+  }
+
+  void reset(rcbuf_t* ptr = nullptr, bool internalizable = false) {
+    this->~NgRcBufPointer();
```

Comment:
I think I’d prefer moving the destructor’s contents manually to here, and then just calling `reset()` from the destructor – it should do the same thing, but looks a bit more correct?